Mm 42023 add task list to end users

pvev opened the pull request:
* Note: This involves a lot of code removal, this will need extra check and validation to make sure of two things:
- By removing the former "Next Steps" we didn't break any corner case functionality
- The new onboarding task list works as expected for First Admins, Admins and End users.

#### Summary
This PR removes the Next Steps screen and feature and enables the new onboarding task list for admins (non first admins) and end users (those with a restriction for seeing the options "start trial" and "visit system console")

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-42023

#### Related Pull Requests
n/a

#### Screenshots


#### Release Note
```release-note
This PR enables the new onboarding task list to admins (non first admins) and end users
```

pvev commented:
adding @stevemudie to this draft for early QA feedback.

pvev commented:
/e2e-test

mattermod commented:
Successfully triggered e2e testing!
https://git.internal.mattermost.com/qa/cypress-ui-automation/-/pipelines/183014

neallred reviewed (changes requested):
That's a lot of code deletions :heart_eyes: ! Besides the few issues commented below, I noticed that locally as a non-admin user, I could complete the channels tour, but the item didn't get crossed off the onboarding task list, and when I clicked on the item again, I was again started over in the channels tour.

pvev commented on components/root/index.js:
```diff
@@ -31,9 +31,10 @@ function mapStateToProps(state) {
-    const showTaskList = isUserFirstAdmin && taskListStatus && !isMobileView;
+    const showTaskList = true || (taskListStatus && !isMobileView);
+
+    // send a value to verify is admin so the two new values of the list are shown
```
sorry, a left over of the latest local tests 🤦

pvev commented on components/onboarding_tasks/utils.ts:
```diff
@@ -4,3 +4,7 @@
 };
+
+export function getAnalyticsCategory(isAdmin: boolean) {
+    return isAdmin ? 'cloud_first_user_onboarding' : 'cloud_end_user_onboarding';
```
right!

pvev commented on components/sidebar/sidebar_channel/sidebar_channel_link/index.ts:
```diff
@@ -44,12 +43,10 @@ function makeMapStateToProps() {
         const isUserFirstAdmin = isFirstAdmin(state);
         const showChannelsTour = enableTutorial && tutorialStep === OnboardingTourSteps.CHANNELS_AND_DIRECT_MESSAGES;
-        const showChannelsTutorialStep = showChannelsTour && ((channelTourTriggered && isUserFirstAdmin) || (isOnboardingFlowEnabled === 'true' && !(nextStep || onboardingHidden) && !isUserFirstAdmin) || (isOnboardingFlowEnabled !== 'true' && !isUserFirstAdmin));
+        const showChannelsTutorialStep = showChannelsTour && ((channelTourTriggered && isUserFirstAdmin) || (isOnboardingFlowEnabled === 'true' && !isUserFirstAdmin) || (isOnboardingFlowEnabled !== 'true' && !isUserFirstAdmin));
```
let me verify this one!

pvev replied:
done!

pvev commented on plugins/test/__snapshots__/main_menu_action.test.jsx.snap:
```diff
@@ -446,7 +457,25 @@ exports[`plugins/MainMenuActions should match snapshot in web view 1`] = `
         teamId="someteamid"
-      />
+      >
+        <MenuItemToggleModalRedux
```
mmm, not sure, let me verify that

pvev replied:
no unintended change, just removing previously required props used for the previous next steps. It appeared because I added this value that TS was complaining for it not to be present on the test: https://github.com/mattermost/mattermost-webapp/pull/10187/files#diff-1c5f991b9fcdf25fe33e9a38571dcf9a5ebdd1df67f58dedefa1e2afbb264222R34

pvev commented on components/preparing_workspace/preparing_workspace.tsx:
```diff
@@ -290,16 +289,6 @@ export default function PreparingWorkspace(props: Props) {
-
-                // store the firstChannelName value to redux and in preferences, also set the defaultStep to firstChannelName (-1)
-                dispatch(setFirstChannelName(redirectChannel.name));
-                dispatch(savePreferences(user.id, [firstChannelNamePref, defaultStepPref]));
```
I had my doubts here, but was not sure of the correct behavior of it. Thanks for clarifying it

stevemudie commented:
Hey @pvev I've taken a first pass and the onboarding task list changes after I bump a user up to admin so that's good but no matter what I do the 'Take a tour of channels' step won't show as completed.

pvev commented:
> Hey @pvev I've taken a first pass and the onboarding task list changes after I bump a user up to admin so that's good but no matter what I do the 'Take a tour of channels' step won't show as completed.

hey @stevemudie yes, that seems to be a problem with some code I removed from the first channel creation code. I am working on it now. Will keep you posted once it gets fixed. Thanks!

neallred commented on components/sidebar/sidebar_channel/sidebar_channel_link/index.ts:
```diff
@@ -44,13 +42,9 @@ function makeMapStateToProps() {
         const showChannelsTour = enableTutorial && tutorialStep === OnboardingTourSteps.CHANNELS_AND_DIRECT_MESSAGES;
-        const showChannelsTutorialStep = showChannelsTour && ((channelTourTriggered && isUserFirstAdmin) || (isOnboardingFlowEnabled === 'true' && !(nextStep || onboardingHidden) && !isUserFirstAdmin) || (isOnboardingFlowEnabled !== 'true' && !isUserFirstAdmin));
-
+        const showChannelsTutorialStep = showChannelsTour && channelTourTriggered && isOnboardingFlowEnabled === 'true';
```
I haven't dug in, but it looks like this logic is not enough. I'm guessing there's something from the first admin onboarding that doesn't get set for other users? When I create a new regular user on the spinwick, I'm not getting the task list showing.

pvev replied:
let me check that

marianunez commented on components/onboarding_tasks/onboarding_tasks_manager.tsx:
```diff
@@ -113,6 +116,11 @@ export const useTasksList = () => {
         delete list.START_TRIAL;
     }
 
+    if (!isUserFirstAdmin && !isUserAdmin) {
+        delete list.VISIT_SYSTEM_CONSOLE;
+        delete list.START_TRIAL;
+    }
```
Can we add test cases (either unit or cypress tests) to cover when the task list should show/not show for admin and end users?

neallred commented on components/root/root.jsx:
```diff
@@ -280,6 +283,39 @@ export default class Root extends React.PureComponent {
+        const hasUserStartedOnboarding = preferences.some((pref) =>
+            pref.name === OnboardingTaskList.ONBOARDING_TASK_LIST_INIT && pref.value === 'true');
+
+        if (!hasUserStartedOnboarding) {
```
I think we'll also need to check whether existing users already have any progress in the tour, so that we don't re-open things on them if they've started or finished the tour.

pvev replied:
oh yes, I am checking the former value we used for the next steps too. I just wanted to keep pushing so I can continue validating using the spinwick server.

JulienTant reviewed (approved):
Overall LGTM, the code looks good, I didn't see anything that feels wrong while reading and the feature worked on the spinwick. Now I think @neallred and @AshishDhama reviews will be more useful as they have more knowledge on the feature.

neallred commented on components/root/index.js:
```diff
@@ -23,17 +26,44 @@ import {isMobile} from 'utils/utils.jsx';
+        const taskListStatus = getBool(state, OnboardingTaskCategory, OnboardingTaskList.ONBOARDING_TASK_LIST_SHOW);
+        const isMobileView = isMobile();
+        showTaskList = (firstTimeOnboarding || taskListStatus) && !isMobileView;
+    }
```
Good finds here for preserving existing user experience! :100:

neallred reviewed (approved):
Nice work removing code and with all the follow-ups to make sure we don't double show this to existing users! LGTM, @pvev !

marianunez commented on components/preparing_workspace/preparing_workspace.tsx:
```diff
@@ -364,22 +362,6 @@ export default function PreparingWorkspace(props: Props) {
             }
         };
 
-        // prepare task list to open in the next view
-        dispatch(savePreferences(user.id, [
-            {
-                user_id: user.id,
-                category: OnboardingTaskCategory,
-                name: OnboardingTaskList.ONBOARDING_TASK_LIST_SHOW,
-                value: 'true',
-            },
-            {
-                user_id: user.id,
-                category: OnboardingTaskCategory,
-                name: OnboardingTaskList.ONBOARDING_TASK_LIST_OPEN,
-                value: 'true',
-            },
-        ]));
-
```
Is this no longer needed to open the task list for the first admin?

pvev replied:
no, all users including first admins will get the initial values (show/open) set.

marianunez commented on components/root/index.js:
```diff
@@ -23,17 +26,44 @@ import {isMobile} from 'utils/utils.jsx';
+    let firstTimeOnboarding = false;
+
+    // validation to avoid execute logic on first load which has no preferences values on global store
+    if (Object.keys(state.entities.preferences.myPreferences).length > 0) {
+        // conditions to validate scenario where users (initially first_admins) had already set any of the onboarding task list preferences values.
+        // We check wether the preference value exists meaning the onboarding tasks list already started no matter what the state of the process is
+        const onboardingPreferences = getCategory(state, OnboardingTaskCategory);
+        const hasUserStartedOnboardingTaskListProcess = onboardingPreferences?.some((pref) =>
+            pref.name === OnboardingTaskList.ONBOARDING_TASK_LIST_SHOW || pref.name === OnboardingTaskList.ONBOARDING_TASK_LIST_OPEN);
+
+        // This condition verifies existing users hasn't finished nor skipped legacy next steps or there are still steps not completed
+        const legacyStepsPreferences = getCategory(state, Preferences.RECOMMENDED_NEXT_STEPS);
+        const hasLegacyNextStepsPrefs = legacyStepsPreferences?.some((pref) =>
+            pref.name === RecommendedNextSteps.SKIP || pref.name === RecommendedNextSteps.HIDE);
+
+        const existingUserHasntFinishedNorSkippedLegacyNextSteps = hasLegacyNextStepsPrefs && showOnboardingTaskListToExistingUsers(state);
+
+        const completelyNewUserForOnboarding = !hasUserStartedOnboardingTaskListProcess && !hasLegacyNextStepsPrefs;
+
+        if (completelyNewUserForOnboarding || existingUserHasntFinishedNorSkippedLegacyNextSteps) {
+            firstTimeOnboarding = true;
+        }
+
+        const taskListStatus = getBool(state, OnboardingTaskCategory, OnboardingTaskList.ONBOARDING_TASK_LIST_SHOW);
+        const isMobileView = isMobile();
+        showTaskList = (firstTimeOnboarding || taskListStatus) && !isMobileView;
```
Do we have any test coverage for all these cases? We may have an opportunity to extract this as a function and add some unit test - this is a vital code path that can cause every existing user to get the checklist on upgrade or no one getting the checklist too :D

pvev replied:
correct! this is the piece of code that handles all the logic to determine when to show/hide the task list. I will extract it to a function and write unit test for it.

marianunez commented on selectors/onboarding.ts:
```diff
@@ -34,3 +33,92 @@ export function getFirstChannelName(state: GlobalState) {
+);
+
+// Loop through all Steps. For each step, check that
+export const legacyNextStepsNotFinished = createSelector(
+    'legacyNextStepsNotFinished',
+    (state: GlobalState) => getCategory(state, Preferences.RECOMMENDED_NEXT_STEPS),
+    (state: GlobalState) => getCurrentUser(state),
+    (state: GlobalState) => isFirstAdmin(state),
+    (state: GlobalState) => getSteps(state),
+    (stepPreferences, currentUser, firstAdmin, mySteps) => {
+        const roles = firstAdmin ? `first_admin ${currentUser?.roles}` : currentUser?.roles;
+        const checkPref = (step: StepType) => stepPreferences.some((pref) => (pref.name === step.id && pref.value === 'true') || !isStepForUser(step, roles));
+        return !mySteps.every(checkPref);
+    },
+);
```
Do we need to check for each step still ? It seems that these don't correlate to the new onboarding list

pvev replied:
yes, they don't correlate, but it makes sense that if a user hasn't finish the previous steps to see the new task list. Could potentially help end users to learn new things and admins to get the motivation to start a trial.

marianunez commented on package-lock.json:
```diff
@@ -7,6 +7,9 @@
     "": {
       "name": "@mattermost/webapp",
       "version": "5.36.0-0",
+      "workspaces": [
```
Are the changes in this file intended? It's seems to be conflicting with master

pvev replied:
nope, not intented at all. Will remove this. Thanks, nice catch.

pvev commented on components/preparing_workspace/preparing_workspace.tsx:
```diff
@@ -22,9 +22,7 @@ import {isFirstAdmin} from 'mattermost-redux/selectors/entities/users';
-import {OnboardingTaskCategory, OnboardingTaskList} from 'components/onboarding_tasks/constants';
-
-import Constants, {Preferences, RecommendedNextSteps} from 'utils/constants';
+import Constants, {RecommendedNextSteps, Preferences} from 'utils/constants';
```
hey, yes, we need it for the validations for older preferences values, it is just a placeholder for constant values.

marianunez reviewed (approved):
Great work @pvev 🎉

stevemudie commented:
@pvev This is looking great and worked as expected except for _one_ small element.

The end/regular member sees the `Start free 30-day trial` button at the end of the steps when they should only see the `Got it` button based on [what I'm seeing in Figma](https://www.figma.com/file/LUePg1UMwUmxfxgmT4KCzC/MM-42023-End-User-Onboarding-Task-List?node-id=419%3A120181) anyway.

![Screen Shot 2022-05-13 at 2 48 26 PM](https://user-images.githubusercontent.com/7083912/168360877-cdbab2a9-f354-406b-b355-dbfbd5f87839.png)

pvev commented:
> @pvev This is looking great and worked as expected except for _one_ small element.
> 
> The end/regular member sees the `Start free 30-day trial` button at the end of the steps when they should only see the `Got it` button based on [what I'm seeing in Figma](https://www.figma.com/file/LUePg1UMwUmxfxgmT4KCzC/MM-42023-End-User-Onboarding-Task-List?node-id=419%3A120181) anyway.
> 
> ![Screen Shot 2022-05-13 at 2 48 26 PM](https://user-images.githubusercontent.com/7083912/168360877-cdbab2a9-f354-406b-b355-dbfbd5f87839.png)

Fixed @stevemudie 

Completed tasks for:
Normal End user:
<img width="1781" alt="Captura de pantalla 2022-05-14 a las 1 36 25" src="https://user-images.githubusercontent.com/10082627/168401612-c9b72cf9-bea9-4a45-aa78-6b9ff09543f7.png">

First Admin/Admin users:
<img width="1290" alt="Captura de pantalla 2022-05-14 a las 1 36 44" src="https://user-images.githubusercontent.com/10082627/168401647-6f8c551c-5837-4c00-910d-745445939262.png">

stevemudie commented:
/e2e-test

mattermod commented:
Successfully triggered e2e testing!
https://git.internal.mattermost.com/qa/cypress-ui-automation/-/pipelines/194923

pvev commented:
/e2e-test

mattermod commented:
Successfully triggered e2e testing!
https://git.internal.mattermost.com/qa/cypress-ui-automation/-/pipelines/195098

pvev commented:
/e2e-test

mattermod commented:
Successfully triggered e2e testing!
https://git.internal.mattermost.com/qa/cypress-ui-automation/-/pipelines/195198

Phrynobatrachus commented:
/e2e-test

pvev commented:
/e2e-test

mattermod commented:
Successfully triggered e2e testing!
https://git.internal.mattermost.com/qa/cypress-ui-automation/-/pipelines/195499

saturninoabril commented:
/e2e-test

mattermod commented:
Successfully triggered e2e testing!
https://git.internal.mattermost.com/qa/cypress-ui-automation/-/pipelines/195672

pvev commented:
@Phrynobatrachus can you please also help with the e2e test rerun here and also review the latest changes I added to the cypress tests? 🙏

Phrynobatrachus commented:
/e2e-test

Phrynobatrachus commented:
/e2e-test

mattermod commented:
Successfully triggered e2e testing!
https://git.internal.mattermost.com/qa/cypress-ui-automation/-/pipelines/195942

Phrynobatrachus commented:
The problem with the previous commit was that the newly created admin wasn't logged prior to make the preferences API calls.

Phrynobatrachus commented:
/e2e-test

pvev commented:
/e2e-test

mattermod commented:
Successfully triggered e2e testing!
https://git.internal.mattermost.com/qa/cypress-ui-automation/-/pipelines/196184

pvev commented:
hi @Phrynobatrachus , I added the fixes for the onboarding tests and removed those that were no longer relevant cause were testing the deprecated nextSteps feature. This is the commit https://github.com/mattermost/mattermost-webapp/pull/10187/commits/d61d6d11ec9cce208cf96f8bf6fe86a6e8038af7 . I also triggered the e2e but not sure if they run correctly. Thanks!

mattermod commented:
Successfully triggered e2e testing!
https://git.internal.mattermost.com/qa/cypress-ui-automation/-/pipelines/196321

Phrynobatrachus reviewed (approved):
https://automation-dashboard.vercel.app/cycles/f8410bfb-84c2-43d9-8df9-dc0c79c8a8cd

Latest run :+1:, remaining few failures should be unrelated.

pvev commented:
> https://automation-dashboard.vercel.app/cycles/f8410bfb-84c2-43d9-8df9-dc0c79c8a8cd
> 
> Latest run 👍, remaining few failures should be unrelated.

awesome! thanks @Phrynobatrachus .  
@stevemudie e2e fixed! 🚢

marianunez commented:
> https://automation-dashboard.vercel.app/cycles/f8410bfb-84c2-43d9-8df9-dc0c79c8a8cd
> 
> Latest run 👍, remaining few failures should be unrelated.

Thanks guys, great work battling the E2E tests here @Phrynobatrachus @pvev 🎉 💯

stevemudie reviewed (approved):
Looks good @pvev Thank you for your patience and attention to the E2E!

marianunez commented:
@amyblais we would like to cherry pick to `cloud` as the changes here are needed to run an A/B test before Freemium launch. Let us know if you have any concerns

amyblais commented:
/cherry-pick cloud

mattermod commented:
Cherry pick is scheduled.
